src/contracts/project.cairo

```diff
@@ -358,13 +377,183 @@ mod Project {
+        }
+
+        fn internal_to_cc(
+            self: @ContractState, internal_value_to_send: u256, token_id: u256
+        ) -> u256 {
+            let vintage_supply = self.vintage.get_carbon_vintage(token_id).supply.into();
+            let initial_project_supply = self.vintage.get_initial_project_cc_supply();
+            internal_value_to_send * vintage_supply / initial_project_supply
+        }
     }
 
     #[generate_trait]
     impl InternalImpl of InternalTrait {
         fn _balance_of(self: @ContractState, account: ContractAddress, token_id: u256) -> u256 {
-            let share = self.erc1155.balance_of(account, token_id);
-            self.vintage.share_to_cc(share, token_id)
+            self.internal_to_cc(self.erc1155.balance_of(account, token_id), token_id)
+        }
+
+        fn _mint(ref self: ContractState, to: ContractAddress, token_id: u256, value: u256) {
+            let isMinter = self.accesscontrol.has_role(MINTER_ROLE, get_caller_address());
+            assert(isMinter, 'Only Minter can mint');
+            self.erc1155.mint(to, token_id, value);
+            let cc_value = self.internal_to_cc(value, token_id);
+            self
+                .emit(
+                    ERC1155Component::Event::TransferSingle(
+                        ERC1155Component::TransferSingle {
+                            operator: get_contract_address(),
+                            from: get_contract_address(),
+                            to,
+                            id: token_id,
+                            value: cc_value,
+                        }
+                    )
+                );
+        }
+
+        fn _batch_mint(
+            ref self: ContractState, to: ContractAddress, token_ids: Span<u256>, values: Span<u256>
+        ) {
+            self.erc1155.batch_mint(to, token_ids, values);
+            let mut values_to_emit: Array<u256> = Default::default();
+            let self_snap = @self;
+            let mut index = 0;
+            loop {
+                if index == token_ids.len() {
+                    break;
+                }
+                let cc_value = self_snap.internal_to_cc(*values.at(index), *token_ids.at(index));
+                values_to_emit.append(cc_value);
+                index += 1;
+            };
+            let values_to_emit = values_to_emit.span();
+
+            self
+                .emit(
+                    ERC1155Component::Event::TransferBatch(
+                        ERC1155Component::TransferBatch {
+                            operator: get_caller_address(),
+                            from: Zeroable::zero(),
+                            to,
+                            ids: token_ids,
+                            values: values_to_emit,
+                        }
+                    )
+                );
+        }
+
+        fn _offset(ref self: ContractState, from: ContractAddress, token_id: u256, value: u256) {
```

Comment:
I think the name `burn` should be kept. As the method is actually burning the token while offsetting requires more stuff to be done (by the offsetter contract).